File: mobile/lib/src/features/auth/data/social_auth_service.dart
import 'dart:async';

import 'package:google_sign_in/google_sign_in.dart';
import 'package:sign_in_with_apple/sign_in_with_apple.dart';

import '../../../core/api/dio_client.dart';
import '../domain/user.dart';

/// Result of social authentication
class SocialAuthResult {
  final User user;
  final String token;
  final String refreshToken;
  final bool isNewUser;

  SocialAuthResult({
    required this.user,
    required this.token,
    required this.refreshToken,
    required this.isNewUser,
  });

  factory SocialAuthResult.fromJson(Map<String, dynamic> json) {
    return SocialAuthResult(
      user: User.fromJson(json['user'] as Map<String, dynamic>),
      token: json['token'] as String,
      refreshToken: json['refreshToken'] as String,
      isNewUser: json['isNewUser'] as bool? ?? false,
    );
  }
}

class AppleSocialCredential {
  const AppleSocialCredential({
    required this.identityToken,
    this.givenName,
    this.familyName,
  });

  final String identityToken;
  final String? givenName;
  final String? familyName;
}

/// Low-level platform boundary for the Google and Apple SDKs.
abstract interface class SocialAuthPlatform {
  Future<String?> signInWithGoogle();
  Future<AppleSocialCredential> signInWithApple();
  Future<void> signOutGoogle();
}

class DefaultSocialAuthPlatform implements SocialAuthPlatform {
  final GoogleSignIn _googleSignIn = GoogleSignIn.instance;
  Completer<void>? _initCompleter;

  Future<void> _ensureGoogleSignInInitialized() async {
    if (_initCompleter != null) {
      return _initCompleter!.future;
    }
    _initCompleter = Completer<void>();
    try {
      await _googleSignIn.initialize();
      _initCompleter!.complete();
    } catch (_) {
      _initCompleter = null;
      rethrow;
    }
  }

  @override
  Future<String?> signInWithGoogle() async {
    await _ensureGoogleSignInInitialized();
    var account = await _googleSignIn.attemptLightweightAuthentication();
    if (account == null && _googleSignIn.supportsAuthenticate()) {
      account = await _googleSignIn.authenticate();
    }
    if (account == null) return null;

    final idToken = account.authentication.idToken;
    if (idToken == null || idToken.isEmpty) {
      throw Exception('Failed to get Google ID token');
    }
    return idToken;
  }

  @override
  Future<AppleSocialCredential> signInWithApple() async {
    final credential = await SignInWithApple.getAppleIDCredential(
      scopes: [
        AppleIDAuthorizationScopes.email,
        AppleIDAuthorizationScopes.fullName,
      ],
    );
    final identityToken = credential.identityToken;
    if (identityToken == null || identityToken.isEmpty) {
      throw Exception('Failed to get Apple identity token');
    }
    return AppleSocialCredential(
      identityToken: identityToken,
      givenName: credential.givenName,
      familyName: credential.familyName,
    );
  }

  @override
  Future<void> signOutGoogle() async {
    await _ensureGoogleSignInInitialized();
    await _googleSignIn.disconnect();
  }
}

/// Service for handling social authentication (Google, Apple).
///
/// This service handles the client-side OAuth flow and sends tokens
/// to the backend for verification and account management.
class SocialAuthService {
  final DioClient _dioClient;
  final SocialAuthPlatform _platform;

  SocialAuthService({
    required DioClient dioClient,
    SocialAuthPlatform? platform,
  }) : _dioClient = dioClient,
       _platform = platform ?? DefaultSocialAuthPlatform();

  /// Initialize Google Sign-In (required for google_sign_in 7.x)
  /// Uses Completer pattern to avoid race conditions on concurrent calls.
  Future<String> _fetchOAuthState() async {
    final response = await _dioClient.get('/auth/social/state');
    final data = response.data['data'] as Map<String, dynamic>?;
    final state = data?['state'] as String?;
    if (state == null || state.length != 64) {
      throw Exception('Invalid OAuth state from server');
    }
    return state;
  }

  /// Sign in with Google.
  ///
  /// Gets the Google ID token from the client and sends it to the backend
  /// for verification. The backend handles user creation/linking and
  /// returns app-specific auth tokens.
  ///
  /// Returns [SocialAuthResult] with user data and tokens,
  /// or null if the user cancelled the sign-in flow.
  Future<SocialAuthResult?> signInWithGoogle() async {
    final idToken = await _platform.signInWithGoogle();
    if (idToken == null) return null;

    final state = await _fetchOAuthState();

    final response = await _dioClient.post(
      '/auth/social/google',
      data: {'idToken': idToken, 'state': state},
    );

    final data = response.data['data'] as Map<String, dynamic>;
    return SocialAuthResult.fromJson(data);
  }

  /// Sign out from Google.
  Future<void> signOutGoogle() async {
    await _platform.signOutGoogle();
  }

  /// Sign in with Apple.
  ///
  /// Gets the Apple identity token from the client and sends it to the backend
  /// for verification. The backend handles user creation/linking and
  /// returns app-specific auth tokens.
  ///
  /// Note: Apple only provides email/name on first authorization.
  /// Subsequent sign-ins return empty values for these fields.
  ///
  /// Returns [SocialAuthResult] with user data and tokens,
  /// or null if the user cancelled the sign-in flow.
  Future<SocialAuthResult?> signInWithApple() async {
    final credential = await _platform.signInWithApple();

    final requestData = <String, dynamic>{
      'identityToken': credential.identityToken,
    };

    if (credential.givenName != null || credential.familyName != null) {
      requestData['fullName'] = {
        'givenName': credential.givenName,
        'familyName': credential.familyName,
      };
    }

    final state = await _fetchOAuthState();
    requestData['state'] = state;

    final response = await _dioClient.post(
      '/auth/social/apple',
      data: requestData,
    );

    final data = response.data['data'] as Map<String, dynamic>;
    return SocialAuthResult.fromJson(data);
  }
}

File: backend/src/services/SocialAuthService.ts
import { OAuth2Client } from 'google-auth-library';
import appleSignin from 'apple-signin-auth';
import Database from '../config/database';
import { User, AuthResponse } from '../types';
import { AuthUtils } from '../utils/auth';
import { generateRefreshToken } from '../utils/auth';
import { mapDbUserToUser, sanitizeUserForClient } from '../utils/dbMappers';
import { PoolClient } from 'pg';
import crypto from 'crypto';
import logger from '../utils/logger';
import { getRedis } from '../utils/redisRateLimiter';

/**
 * OAuth state validation error
 */
export class OAuthStateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'OAuthStateError';
  }
}

/**
 * Profile information extracted from social auth tokens
 */
export interface SocialProfile {
  provider: 'google' | 'apple';
  providerId: string;
  email: string;
  firstName?: string;
  lastName?: string;
  profileImageUrl?: string;
}

/**
 * Result of social authentication
 */
export interface SocialAuthResult extends AuthResponse {
  refreshToken: string;
  isNewUser: boolean;
}

/**
 * Service for handling social authentication (Google, Apple)
 * Verifies tokens server-side and creates/links user accounts
 */
export class SocialAuthService {
  private db = Database.getInstance();
  private googleClient: OAuth2Client;

  constructor() {
    // Initialize Google OAuth2 client
    // GOOGLE_CLIENT_ID should be set in environment variables
    const googleClientId = process.env.GOOGLE_CLIENT_ID;
    this.googleClient = new OAuth2Client(googleClientId);
  }

  /**
   * Generate OAuth state parameter for CSRF protection
   * Stores state in Redis with 5-minute TTL
   * @returns The generated state token
   */
  async generateOAuthState(): Promise<string> {
    const redis = getRedis();
    if (!redis) {
      logger.warn('Redis unavailable, cannot generate OAuth state (CSRF protection disabled)');
      throw new OAuthStateError('OAuth state generation requires Redis');
    }

    const state = crypto.randomBytes(32).toString('hex');
    // Store in Redis with short TTL (5 minutes)
    await redis.setex(`oauth:state:${state}`, 300, 'valid');
    return state;
  }

  /**
   * Validate OAuth state parameter to prevent CSRF attacks
   * One-time use - deletes state from Redis after validation
   * @param state - The state parameter from OAuth callback
   * @returns true if state is valid, false otherwise
   */
  async validateOAuthState(state: string): Promise<boolean> {
    const redis = getRedis();
    if (!redis) {
      logger.warn('Redis unavailable, cannot validate OAuth state');
      return false;
    }

    const key = `oauth:state:${state}`;
    const valid = await redis.get(key);
    if (valid) {
      await redis.del(key); // One-time use
      return true;
    }
    return false;
  }

  /**
   * Verify a Google ID token and extract user profile
   * @param idToken - The Google ID token from the mobile client
   * @param state - Optional OAuth state parameter for CSRF protection
   * @returns The verified user profile or null if invalid
   */
  async verifyGoogleToken(idToken: string, state: string): Promise<SocialProfile | null> {
    try {
      const isValidState = await this.validateOAuthState(state);
      if (!isValidState) {
        logger.error('Invalid OAuth state - possible CSRF attack', { provider: 'google' });
        throw new OAuthStateError('Invalid OAuth state - possible CSRF attack');
      }

      const googleClientId = process.env.GOOGLE_CLIENT_ID;
      if (!googleClientId) {
        logger.error('GOOGLE_CLIENT_ID not configured');
        return null;
      }

      const ticket = await this.googleClient.verifyIdToken({
        idToken,
        audience: googleClientId,
      });

      const payload = ticket.getPayload();
      if (!payload) {
        return null;
      }

      // Ensure email is verified
      if (!payload.email_verified) {
        logger.error('Google email not verified');
        return null;
      }

      return {
        provider: 'google',
        providerId: payload.sub, // Google's unique user ID
        email: payload.email!,
        firstName: payload.given_name,
        lastName: payload.family_name,
        profileImageUrl: payload.picture,
      };
    } catch (error) {
      logger.error('Google token verification failed', {
        error: error instanceof Error ? error.message : String(error),
        stack: error instanceof Error ? error.stack : undefined,
      });
      throw error;
    }
  }

  /**
   * Verify an Apple identity token and extract user profile
   * Uses apple-signin-auth to properly verify the token signature against Apple's public keys
   * @param identityToken - The Apple identity token from the mobile client
   * @param fullName - Optional full name (only provided on first sign-in)
   * @param state - Optional OAuth state parameter for CSRF protection
   * @returns The verified user profile or null if invalid
   */
  async verifyAppleToken(
    identityToken: string,
    fullName: { givenName?: string; familyName?: string } | undefined,
    state: string
  ): Promise<SocialProfile | null> {
    try {
      const isValidState = await this.validateOAuthState(state);
      if (!isValidState) {
        logger.error('Invalid OAuth state - possible CSRF attack', { provider: 'apple' });
        throw new OAuthStateError('Invalid OAuth state - possible CSRF attack');
      }

      // Require APPLE_BUNDLE_ID for Apple sign-in
      const appleBundleId = process.env.APPLE_BUNDLE_ID;
      if (!appleBundleId) {
        logger.error('APPLE_BUNDLE_ID environment variable required for Apple sign-in');
        throw new Error('APPLE_BUNDLE_ID environment variable required for Apple sign-in');
      }

      // Verify the token signature against Apple's public keys
      const appleUser = await appleSignin.verifyIdToken(identityToken, {
        audience: appleBundleId,
        ignoreExpiration: false,
      });

      if (!appleUser.sub) {
        logger.error('Invalid Apple token: missing subject');
        return null;
      }

      // Email might not be present after first sign-in
      // We need to look it up from existing social account if not provided
      const email = appleUser.email;

      return {
        provider: 'apple',
        providerId: appleUser.sub,
        email: email || '', // May be empty, will be handled in authenticateOrCreate
        firstName: fullName?.givenName,
        lastName: fullName?.familyName,
      };
    } catch (error: any) {
      logger.error('Apple token verification failed', {
        error: error instanceof Error ? error.message : String(error),
        stack: error instanceof Error ? error.stack : undefined,
      });
      throw error;
    }
  }

  /**
   * Authenticate or create a user based on social profile
   * - If social account exists, return existing user
   * - If email exists but not linked, link social account to existing user
   * - If neither exists, create new user with unique username
   *
   * @param profile - The verified social profile
   * @returns Authentication result with user, token, and whether user is new
   */
  async authenticateOrCreate(profile: SocialProfile): Promise<SocialAuthResult> {
    // First, check if this social account already exists
    const existingLink = await this.findSocialAccount(profile.provider, profile.providerId);

    if (existingLink) {
      // User already has this social account linked - return existing user
      const user = await this.findUserById(existingLink.userId);
      if (!user) {
        throw new Error('User not found for social account');
      }

      if (!user.isActive) {
        throw new Error('Account is deactivated');
      }

      return this.generateAuthResult(user, false);
    }

    // Social account doesn't exist - check if we have an email to work with
    if (!profile.email) {
      // For Apple re-auth: email may be empty after first sign-in.
      // Try to find an existing social account by provider + providerId
      // (the initial findSocialAccount above checks provider_id, but the link
      // may not exist yet if the user was created via email and this is a
      // first-time Apple sign-in with no email provided).
      // If we truly have no email and no existing link, we cannot proceed.
      throw new Error('Email is required for new social sign-in');
    }

    // Check if a user with this email already exists
    const existingUser = await this.findUserByEmail(profile.email);

    if (existingUser) {
      // Check if this is a password-only account (no social accounts linked).
      // If so, don't auto-link — this prevents account takeover via social auth.
      // The user must log in with their password first and link the social account
      // from account settings.
      const hasSocialAccounts = await this.userHasSocialAccounts(existingUser.id);
      if (!hasSocialAccounts) {
        logger.warn('Social auth email matches password-only account, refusing auto-link', {
          provider: profile.provider,
          existingUserId: existingUser.id,
        });
        const err = new Error(
          'An account with this email already exists. Please log in with your password and link your social account from settings.'
        );
        (err as any).statusCode = 409;
        throw err;
      }

      // User already has social auth — safe to link this additional provider
      await this.linkSocialAccount(existingUser.id, profile.provider, profile.providerId);
      return this.generateAuthResult(existingUser, false);
    }

    // No existing user - create a new account
    const newUser = await this.createSocialUser(profile);
    return this.generateAuthResult(newUser, true);
  }

  /**
   * Find a social account by provider and provider ID
   */
  private async findSocialAccount(
    provider: string,
    providerId: string
  ): Promise<{ userId: string } | null> {
    const result = await this.db.query(
      `SELECT user_id FROM user_social_accounts
       WHERE provider = $1 AND provider_id = $2`,
      [provider, providerId]
    );

    if (result.rows.length === 0) {
      return null;
    }

    return { userId: result.rows[0].user_id };
  }

  /**
   * Find a user by ID
   */
  private async findUserById(userId: string): Promise<User | null> {
    const result = await this.db.query(
      `SELECT id, email, username, first_name, last_name, bio, profile_image_url,
              location, date_of_birth, is_verified, is_active, created_at, updated_at
       FROM users WHERE id = $1`,
      [userId]
    );

    if (result.rows.length === 0) {
      return null;
    }

    return mapDbUserToUser(result.rows[0]);
  }

  /**
   * Find a user by email
   */
  private async findUserByEmail(email: string): Promise<User | null> {
    const result = await this.db.query(
      `SELECT id, email, username, first_name, last_name, bio, profile_image_url,
              location, date_of_birth, is_verified, is_active, created_at, updated_at
       FROM users WHERE email = $1 AND is_active = true`,
      [email.toLowerCase()]
    );

    if (result.rows.length === 0) {
      return null;
    }

    return mapDbUserToUser(result.rows[0]);
  }

  /**
   * Check if a user has any social accounts linked.
   * Used to distinguish password-only users from social-auth users.
   */
  private async userHasSocialAccounts(userId: string): Promise<boolean> {
    const result = await this.db.query(
      `SELECT 1 FROM user_social_accounts WHERE user_id = $1 LIMIT 1`,
      [userId]
    );
    return result.rows.length > 0;
  }

  /**
   * Link a social account to an existing user
   */
  private async linkSocialAccount(
    userId: string,
    provider: string,
    providerId: string
  ): Promise<void> {
    await this.db.query(
      `INSERT INTO user_social_accounts (user_id, provider, provider_id)
       VALUES ($1, $2, $3)
       ON CONFLICT (provider, provider_id) DO NOTHING`,
      [userId, provider, providerId]
    );
  }
  /**
   * Link a social account using a specific database client (for transactions)
   */
  private async linkSocialAccountWithClient(
    client: PoolClient,
    userId: string,
    provider: string,
    providerId: string
  ): Promise<void> {
    await client.query(
      `INSERT INTO user_social_accounts (user_id, provider, provider_id)
       VALUES ($1, $2, $3)
       ON CONFLICT (provider, provider_id) DO NOTHING`,
      [userId, provider, providerId]
    );
  }

  /**
   * Create a new user from social profile
   * Uses a database transaction to ensure user creation and social account linking are atomic
   */
  private async createSocialUser(profile: SocialProfile): Promise<User> {
    const client = await this.db.getPool().connect();
    try {
      await client.query('BEGIN');

      // Generate a unique username based on email or name
      const baseUsername = this.generateBaseUsername(profile);
      const username = await this.generateUniqueUsernameWithClient(client, baseUsername);

      // Generate random password hash for social auth users (not guessable, not plaintext)
      const randomPassword = crypto.randomBytes(32).toString('hex');
      const hashedPlaceholder = await AuthUtils.hashPassword(randomPassword);

      // Create user with random bcrypt hash (social auth users authenticate via provider)
      const result = await client.query(
        `INSERT INTO users (email, password_hash, username, first_name, last_name, profile_image_url, is_verified)
         VALUES ($1, $2, $3, $4, $5, $6, true)
         RETURNING id, email, username, first_name, last_name, bio, profile_image_url,
                   location, date_of_birth, is_verified, is_active, created_at, updated_at`,
        [
          profile.email.toLowerCase(),
          hashedPlaceholder, // Random bcrypt hash (social auth users authenticate via provider)
          username,
          profile.firstName || null,
          profile.lastName || null,
          profile.profileImageUrl || null,
        ]
      );

      const user = mapDbUserToUser(result.rows[0]);

      // Link the social account within the same transaction
      await this.linkSocialAccountWithClient(client, user.id, profile.provider, profile.providerId);

      await client.query('COMMIT');
      return user;
    } catch (error) {
      await client.query('ROLLBACK');
      throw error;
    } finally {
      client.release();
    }
  }
  /**
   * Generate a base username from profile
   */
  private generateBaseUsername(profile: SocialProfile): string {
    // Try to create username from name or email
    if (profile.firstName && profile.lastName) {
      return `${profile.firstName}${profile.lastName}`.toLowerCase().replace(/[^a-z0-9]/g, '');
    }
    if (profile.firstName) {
      return profile.firstName.toLowerCase().replace(/[^a-z0-9]/g, '');
    }
    // Use email prefix as fallback
    return profile.email
      .split('@')[0]
      .toLowerCase()
      .replace(/[^a-z0-9]/g, '');
  }

  /**
   * Generate a unique username using a specific database client (for transactions)
   */
  private async generateUniqueUsernameWithClient(
    client: PoolClient,
    baseUsername: string
  ): Promise<string> {
    // Ensure base username is at least 3 characters
    if (baseUsername.length < 3) {
      baseUsername = baseUsername + 'user';
    }

    // Truncate to leave room for suffix
    if (baseUsername.length > 25) {
      baseUsername = baseUsername.substring(0, 25);
    }

    let username = baseUsername;
    let suffix = 1;

    while (await this.usernameExistsWithClient(client, username)) {
      username = `${baseUsername}${suffix}`;
      suffix++;

      // Safety limit to prevent infinite loop
      if (suffix > 9999) {
        username = `${baseUsername}${Date.now()}`;
        break;
      }
    }

    return username;
  }

  /**
   * Check if a username already exists using a specific database client (for transactions)
   */
  private async usernameExistsWithClient(client: PoolClient, username: string): Promise<boolean> {
    const result = await client.query(`SELECT 1 FROM users WHERE username = $1 LIMIT 1`, [
      username,
    ]);
    return result.rows.length > 0;
  }

  /**
   * Generate authentication result with tokens
   */
  private async generateAuthResult(user: User, isNewUser: boolean): Promise<SocialAuthResult> {
    const token = AuthUtils.generateToken({
      userId: user.id,
      email: user.email,
      username: user.username,
    });

    const refreshToken = await generateRefreshToken(user.id);

    return {
      user: sanitizeUserForClient(user) as User,
      token,
      refreshToken,
      isNewUser,
    };
  }
}
